### PR TITLE
Support new features of Pkl 0.26

### PR DESCRIPTION
**Pull request by bioball**
This supports new message passing options available in Pkl 0.26.

Also:
* Return error if underlying Pkl doesn't support the specified option
* Add (private) semver library to compare Pkl versions
* Adjust CI to test against both Pkl 0.25 and 0.26

### CODE REVIEW

File: pkl/version.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	if a < b {
+		return -1
+	}
+	return 1
+}
+
+func (s *semver) compareToString(other string) int {
+	otherVersion, err := parseSemver(other)
+	if err != nil {
+		return 0
```
Comment (stackoverflow) [inline pkl/version.go:35]: Returning `0` (equals) for bogus versions seems prone to bugs. Would be better to return `int, err` and let the caller check for errors if they want. You could still ignore the error if you don't care.

File: pkl/version.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	if comparison != 0 {
+		return comparison
+	}
+	comparison = compareInt(s.minor, other.minor)
+	if comparison != 0 {
+		return comparison
+	}
+	// technically we should proceed to comparing prerelease versions, but we can skip
+	// this part because we don't have a use-case for it.
+	return compareInt(s.patch, other.patch)
```
Comment (stackoverflow) [inline pkl/version.go:52]: nit: we may not need it for now, but it wouldn't hurt to add a `strings.Compare()` for the `prerelease` and `build`, making the code future-proof.

File: pkl/version.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	parsed, err := parseSemver(s)
+	if err != nil {
+		panic(err)
+	}
+	return parsed
+}
+
+func parseSemver(s string) (*semver, error) {
+	matched := semverPattern.FindStringSubmatch(s)
+	if len(matched) < 5 {
```
Comment (stackoverflow) [inline pkl/version.go:85]:
```suggestion
	if len(matched) < 6 {
```
In order for line 105 not throw you need the array to have at least 6 elements. Or I'm misunderstanding how array indexes work in Go.

File: pkl/version.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+func mustParseSemver(s string) *semver {
+	parsed, err := parseSemver(s)
+	if err != nil {
+		panic(err)
+	}
+	return parsed
+}
+
+func parseSemver(s string) (*semver, error) {
+	matched := semverPattern.FindStringSubmatch(s)
```
Comment (stackoverflow) [inline pkl/version.go:84]: Where is `semverPattern` defined?

**Review by stackoverflow — COMMENTED** [review]
Left some comments.

**Review by holzensp — CHANGES_REQUESTED** [review]

File: pkl/version.go (hunk 2 above)
Comment (bioball) [reply pkl/version.go:52]: Good idea to compare prerelease too.

Build metadata should be excluded from comparison as per the semver spec: https://semver.org/#spec-item-10

File: pkl/version.go (hunk 3 above)
Comment (bioball) [reply pkl/version.go:85]: Yup, you're right!

File: .circleci/config.yml [hunk 5]
```diff
@@ -83,7 +103,7 @@ jobs:
         paths:
         - out/
     docker:
-    - image: cimg/go:1.22
+    - image: cimg/go:1.21
```
[the remark this answers is not in the document]
Comment (bioball) [reply .circleci/config.yml:106]: Yes; two of the jobs were already downgraded to 1.21 in [a previous PR](https://github.com/apple/pkl-go/pull/15/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R25). I don't have the context for why this happened, but regardless our Pkl and YAML were out of sync after that PR was merged.

File: pkl/evaluator_options.go [hunk 6]
```diff
@@ -228,13 +306,17 @@ var WithOsEnv = func(opts *EvaluatorOptions) {
...
 
-func buildEvaluatorOptions(fns ...func(*EvaluatorOptions)) *EvaluatorOptions {
+func buildEvaluatorOptions(version string, fns ...func(*EvaluatorOptions)) *EvaluatorOptions {
 	o := &EvaluatorOptions{}
 	for _, f := range fns {
 		f(o)
 	}
 	// repl:text is the URI of the module used to hold expressions. It should always be allowed.
 	o.AllowedModules = append(o.AllowedModules, "repl:text")
+	if o.Http != nil && pklVersion0_26.isGreaterThanString(version) {
```
[the remark this answers is not in the document]
Comment (bioball) [reply pkl/evaluator_options.go:316]: Yeah, you're right--returning an error makes more sense here.

File: pkl/version.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bioball) [reply pkl/version.go:35]: FYI: version is a private struct and not part of the pkl-go API. This was written pretty pragmatically, but, I just updated this to be a little more principled and return an error if the string is not a semver.

**Review by stackoverflow — APPROVED** [review]
LGTM

File: pkl/version.go [hunk 7]
```diff
@@ -2,21 +2,67 @@ package pkl
...
+	alphaNumericId string
+}
+
+func (i prereleaseIdentifier) compareTo(other prereleaseIdentifier) int {
+	if i.alphaNumericId != "" {
+		if other.alphaNumericId != "" {
+			return strings.Compare(i.alphaNumericId, other.alphaNumericId)
+		} else {
+			return 1
+		}
```
Comment (holzensp) [inline pkl/version.go:40]: This _is_ the behaviour of `strings.Compare`, no?
```suggestion
		return strings.Compare(i.alphaNumericId, other.alphaNumericId)
```

File: pkl/version.go [hunk 8]
```diff
@@ -2,21 +2,67 @@ package pkl
...
+	}
+}
+
+func (s *semver) getPrereleaseIdentifiers() []prereleaseIdentifier {
+	if s.prerelease != "" && len(s.prereleaseIdentifiers) > 0 {
+		return s.prereleaseIdentifiers
+	}
+	if s.prerelease == "" {
+		return nil
+	}
```
Comment (holzensp) [inline pkl/version.go:52]: nit (fewer mental booleans is More Better™)
```suggestion
	if s.prerelease == "" {
		return nil
	}
	if len(s.prereleaseIdentifiers) > 0 {
		return s.prereleaseIdentifiers
	}
```

File: pkl/version.go [hunk 9]
```diff
@@ -49,27 +95,37 @@ func (s *semver) compareTo(other *semver) int {
 	}
 	// technically we should proceed to comparing prerelease versions, but we can skip
 	// this part because we don't have a use-case for it.
```
Comment (holzensp) [inline pkl/version.go:97]:
```suggestion
```

File: pkl/evaluator_options.go [hunk 10]
```diff
@@ -306,18 +307,18 @@ var WithOsEnv = func(opts *EvaluatorOptions) {
...
+func buildEvaluatorOptions(version *semver, fns ...func(*EvaluatorOptions)) (*EvaluatorOptions, error) {
 	o := &EvaluatorOptions{}
 	for _, f := range fns {
 		f(o)
 	}
 	// repl:text is the URI of the module used to hold expressions. It should always be allowed.
 	o.AllowedModules = append(o.AllowedModules, "repl:text")
-	if o.Http != nil && pklVersion0_26.isGreaterThanString(version) {
+	if o.Http != nil && pklVersion0_26.isGreaterThan(version) {
 		// HTTP is not supported on Pkl 0.25; ignore as per documentation of EvaluatorOptions.Http.
```
Comment (holzensp) [inline pkl/evaluator_options.go:318]:
```suggestion
```

**Review by holzensp — APPROVED** [review]
Some clean-up nits.

File: pkl/version.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+	major                 int
+	minor                 int
+	patch                 int
+	prerelease            string
+	build                 string
+	prereleaseIdentifiers []prereleaseIdentifier
+}
+
+type prereleaseIdentifier struct {
+	numbericId     int
```
Comment (holzensp) [inline pkl/version.go:30]: oh, btw... typo
```suggestion
	numericId     int
```

File: pkl/version.go (hunk 7 above)
Comment (bioball) [reply pkl/version.go:40]: Ah, great catch

File: pkl/version.go (hunk 11 above)
Comment (bioball) [reply pkl/version.go]: Fixed!

**Pull request merged**